// backend/src/git-engine/models/RemoteRepository.ts
import { Commit } from './Commit';
import { Branch } from './Branch';
import { FileTree } from 'shared/src/types';

/**
 * RemoteRepository - Represents a remote Git repository
 * Simulates a remote repository that can be cloned, pushed to, and pulled from
 */
export class RemoteRepository {
  public commits: Map<string, Commit>;
  public branches: Map<string, Branch>;

  constructor(
    public readonly name: string,
    public readonly url: string,
    commits: Commit[] = [],
    branches: Branch[] = []
  ) {
    this.commits = new Map();
    this.branches = new Map();

    // Populate commits map
    commits.forEach(commit => this.commits.set(commit.hash, commit));
    
    // Populate branches map
    branches.forEach(branch => this.branches.set(branch.name, branch));
  }

  /**
   * Get a commit by hash
   */
  getCommit(hash: string): Commit | null {
    return this.commits.get(hash) || null;
  }

  /**
   * Get a branch by name
   */
  getBranch(name: string): Branch | null {
    return this.branches.get(name) || null;
  }

  /**
   * Add a commit to the remote repository
   */
  addCommit(commit: Commit): void {
    this.commits.set(commit.hash, commit);
  }

  /**
   * Add a branch to the remote repository
   */
  addBranch(branch: Branch): void {
    this.branches.set(branch.name, branch);
  }

  /**
   * Update a branch to point to a new commit
   */
  updateBranch(branchName: string, commitHash: string): void {
    const branch = this.branches.get(branchName);
    if (branch) {
      branch.updateCommit(commitHash);
    }
  }

  /**
   * Get all commits as an array
   */
  getCommitsArray(): Commit[] {
    return Array.from(this.commits.values());
  }

  /**
   * Get all branches as an array
   */
  getBranchesArray(): Branch[] {
    return Array.from(this.branches.values());
  }

  /**
   * Create a new remote repository
   */
  static create(name: string, url: string): RemoteRepository {
    return new RemoteRepository(name, url);
  }

  /**
   * Clone this remote repository (create a copy of commits and branches)
   */
  clone(): { commits: Commit[], branches: Branch[] } {
    return {
      commits: this.getCommitsArray(),
      branches: this.getBranchesArray().map(b => Branch.create(b.name, b.commitHash))
    };
  }

  /**
   * Serialize remote repository to JSON
   */
  toJSON() {
    return {
      name: this.name,
      url: this.url,
      commits: this.getCommitsArray().map(c => c.toJSON()),
      branches: this.getBranchesArray().map(b => b.toJSON())
    };
  }

  /**
   * Create a RemoteRepository from JSON
   */
  static fromJSON(json: any): RemoteRepository {
    const commits = json.commits.map((c: any) => 
      new Commit(c.hash, c.message, c.author, new Date(c.timestamp), c.parent, c.tree, c.parents)
    );
    const branches = json.branches.map((b: any) => Branch.fromJSON(b));
    
    return new RemoteRepository(
      json.name,
      json.url,
      commits,
      branches
    );
  }
}
